WalletWasabi/BitcoinCore/Rpc/RpcParser.cs

```diff
@@ -27,53 +28,63 @@ public static RpcPubkeyType ConvertPubkeyType(string pubKeyType)
 
 		public static VerboseBlockInfo ParseVerboseBlockResponse(string getBlockResponse)
 		{
-			var blockInfoJson = JObject.Parse(getBlockResponse);
-			var previousBlockHash = blockInfoJson.Value<string>("previousblockhash");
+			var parsed = JsonDocument.Parse(getBlockResponse).RootElement;
+			if (!parsed.TryGetProperty("result", out JsonElement blockInfoJson))
+			{
+				blockInfoJson = parsed;
+			}
+
+			string? previousBlockHash = null;
+			if (blockInfoJson.TryGetProperty("previousblockhash", out var prevBlockHashJson))
+			{
+				previousBlockHash = prevBlockHashJson.GetString();
+			}
+
 			var transaction = new List<VerboseTransactionInfo>();
 
 			var blockInfo = new VerboseBlockInfo(
-				hash: uint256.Parse(blockInfoJson.Value<string>("hash")),
-				prevBlockHash: previousBlockHash is { } ? uint256.Parse(previousBlockHash) : uint256.Zero,
-				confirmations: blockInfoJson.Value<ulong>("confirmations"),
-				height: blockInfoJson.Value<ulong>("height"),
-				blockTime: Utils.UnixTimeToDateTime(blockInfoJson.Value<uint>("time")),
-				transactions: transaction
-			);
+				hash: uint256.Parse(blockInfoJson.GetProperty("hash").GetString()),
+				prevBlockHash: previousBlockHash is not null ? uint256.Parse(previousBlockHash) : uint256.Zero,
+				confirmations: blockInfoJson.GetProperty("confirmations").GetUInt64(),
+				height: blockInfoJson.GetProperty("height").GetUInt64(),
+				blockTime: Utils.UnixTimeToDateTime(blockInfoJson.GetProperty("time").GetUInt32()),
+				transactions: transaction);
 
-			foreach (var txJson in blockInfoJson["tx"])
+			var array = blockInfoJson.GetProperty("tx").EnumerateArray().ToArray();
```

Comment:
I think that `.ToArray()` can be eliminated too to avoid allocation (`for` -> `foreach`).

Reply:
It needs the index in `var txBlockInfo = new TransactionBlockInfo(blockInfo.Hash, blockInfo.BlockTime, i);`.

Reply:
Yes, but you can simply have `foreach` and increment `i` with `i++` yourself, or am I missing something?

Reply:
```csharp
		public static VerboseBlockInfo ParseVerboseBlockResponse(string getBlockResponse)
		{
			var parsed = JsonDocument.Parse(getBlockResponse).RootElement;
			if (!parsed.TryGetProperty("result", out JsonElement blockInfoJson))
			{
				blockInfoJson = parsed;
			}

			string? previousBlockHash = null;
			if (blockInfoJson.TryGetProperty("previousblockhash", out var prevBlockHashJson))
			{
				previousBlockHash = prevBlockHashJson.GetString();
			}

			var transactions = new List<VerboseTransactionInfo>();

			var blockInfo = new VerboseBlockInfo(
				hash: uint256.Parse(blockInfoJson.GetProperty("hash").GetString()),
				prevBlockHash: previousBlockHash is not null ? uint256.Parse(previousBlockHash) : uint256.Zero,
				confirmations: blockInfoJson.GetProperty("confirmations").GetUInt64(),
				height: blockInfoJson.GetProperty("height").GetUInt64(),
				blockTime: Utils.UnixTimeToDateTime(blockInfoJson.GetProperty("time").GetUInt32()),
				transactions: transactions);

			uint i = 0;
			foreach (JsonElement txJson in blockInfoJson.GetProperty("tx").EnumerateArray())
			{
				var inputs = new List<VerboseInputInfo>();
				var outputs = new List<VerboseOutputInfo>();
				var txBlockInfo = new TransactionBlockInfo(blockInfo.Hash, blockInfo.BlockTime, i);
				var tx = new VerboseTransactionInfo(txBlockInfo, uint256.Parse(txJson.GetProperty("txid").GetString()), inputs, outputs);

				foreach (var txinJson in txJson.GetProperty("vin").EnumerateArray())
				{
					VerboseInputInfo input;
					if (txinJson.TryGetProperty("coinbase", out JsonElement cb))
					{
						input = new VerboseInputInfo(cb.GetString() ?? "");
					}
					else
					{
						var prevOut = txinJson.GetProperty("prevout");
						var scriptPubKey = prevOut.GetProperty("scriptPubKey");
						input = new VerboseInputInfo(
							outPoint: new OutPoint(uint256.Parse(txinJson.GetProperty("txid").GetString()), txinJson.GetProperty("vout").GetUInt32()),
							prevOutput: new VerboseOutputInfo(
								value: Money.Coins(prevOut.GetProperty("value").GetDecimal()),
								scriptPubKey: Script.FromHex(scriptPubKey.GetProperty("hex").GetString()),
								pubkeyType: scriptPubKey.GetProperty("type").GetString()));
					}

					inputs.Add(input);
				}

				foreach (var txoutJson in txJson.GetProperty("vout").EnumerateArray())
				{
					var scriptPubKey = txoutJson.GetProperty("scriptPubKey");
					var output = new VerboseOutputInfo(
						value: Money.Coins(txoutJson.GetProperty("value").GetDecimal()),
						scriptPubKey: Script.FromHex(scriptPubKey.GetProperty("hex").GetString()),
						pubkeyType: scriptPubKey.GetProperty("type").GetString());

					outputs.Add(output);
				}

				transactions.Add(tx);
				i++;
			}

			return blockInfo;
		}
```

This is what I meant.

Reply:
Thanks, I understand now. I see that as a more complex solution.